### PR TITLE
chore: performance monitor

### PR DESCRIPTION
**Pull request by JeevaRamu0104**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added New Analytics section called performance monitor which includes chart for 
Payment Distribution By Connector
Payment Distribution By Payment Method
Connector Wise Payment Failure
Method Wise Payment Failure
Connector + Payment Method Wise Payment Failure
image

## How did you test it?
It can't be tested as of now.Its disabled with feature flag performance_monitor

## Where to test it?
- [ ] INTEG
- [ ] SANDBOX
- [ ] PROD

## Checklist

- [ ] I ran `npm run re:build`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by gitanjli525 — APPROVED** [review]

**Review by sagarnaikjuspay — DISMISSED** [review]

File: src/screens/Analytics/PerformanceMonitor/BarChartPerformance/BarChartPerformance.res [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    | _ => ()
+    }
+  }
+  React.useEffect(() => {
+    if startTimeVal->LogicUtils.isNonEmptyString && endTimeVal->LogicUtils.isNonEmptyString {
+      chartFetch()->ignore
+    }
+    None
+  }, [dimensions])
+  <>
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/BarChartPerformance/BarChartPerformance.res:45]: no need for <>

File: src/screens/Analytics/PerformanceMonitor/BarChartPerformance/BarChartPerformanceUtils.res [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    "series": data.series,
+  }->Identity.genericObjectOrRecordToJson
+
+let getStackedBarData = (~array: array<JSON.t>, ~config: chartDataConfig) => {
+  let {groupByKeys} = config
+  let grouped = PerformanceUtils.getGroupByDataForStatusAndPaymentCount(array, groupByKeys)
+  let keys = grouped->Dict.keysToArray
+  let finalResult = Dict.make()
+  let categories = []
+  let _ = keys->Array.forEach(v => {
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/BarChartPerformance/BarChartPerformanceUtils.res:54]: let _ not needed as it won't return anything

File: src/screens/Analytics/PerformanceMonitor/PieChartPerformance/PieChartPerformance.res [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    | _ => ()
+    }
+  }
+  React.useEffect(() => {
+    if startTimeVal->LogicUtils.isNonEmptyString && endTimeVal->LogicUtils.isNonEmptyString {
+      chartFetch()->ignore
+    }
+    None
+  }, [dimensions])
+  <>
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/PieChartPerformance/PieChartPerformance.res:45]: Same <>

File: src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    ->Array.map(key => {
+      dict->getDictFromJsonObject->getString((key: dimension :> string), "")
+    })
+    ->Array.joinWith(":")
+  key
+}
+
+let getGroupByDataForStatusAndPaymentCount = (array, keys: array<dimension>) => {
+  let result = Dict.make()
+  let _ = array->Array.forEach(entry => {
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res:110]: same _

File: src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        let newConnectorResult = Dict.make()
+        let st = entry->getDictFromJsonObject->getString("status", "")
+        let pc = entry->getDictFromJsonObject->getInt("payment_count", 0)
+        Dict.set(result, key, newConnectorResult)
+        Dict.set(newConnectorResult, st, pc)
+      }
+    | Some(connectorResult) => {
+        let st = entry->getDictFromJsonObject->getString("status", "")
+        let pc = entry->getDictFromJsonObject->getInt("payment_count", 0)
+        let currentCount = Dict.get(connectorResult, st)->Belt.Option.getWithDefault(0)
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res:124]:
```suggestion
        let currentCount = Dict.get(connectorResult, st)->Option.getOr(0)
```

File: src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    filtersDict->Dict.set((filter: dimension :> string), data->JSON.Encode.array)
+  })
+  filtersDict->JSON.Encode.object
+}
+
+let getTimeRange = (startTime, endTime) => {
+  [("startTime", startTime->JSON.Encode.string), ("endTimeVal", endTime->JSON.Encode.string)]
+  ->Dict.fromArray
+  ->JSON.Encode.object
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/PerformanceUtils.res:65]:
```suggestion
  ->getJsonFromArrayOfJson
```

File: src/screens/Analytics/PerformanceMonitor/PerformanceMonitorTypes.res [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  getRequestBody: (
+    ~dimensions: dimensions,
+    ~startTime: string,
+    ~endTime: string,
+    ~metrics: array<metrics>,
+    ~groupBy: array<dimension>,
+    ~filters: array<dimension>,
+    ~customFilter: option<dimension>,
+    ~applyFilterFor: option<array<string>>,
+  ) => RescriptCore.JSON.t,
```
Comment (PritishBudhiraja) [inline src/screens/Analytics/PerformanceMonitor/PerformanceMonitorTypes.res:70]:
```suggestion
  ) => JSON.t,
```

**Review by PritishBudhiraja — APPROVED** [review]

**Review by Riddhiagrawal001 — COMMENTED** [review]
Please add Title and subtitle to the page

File: src/screens/Analytics/PerformanceMonitor/PerformanceMonitor.res [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+      endTime: endTimeVal,
+      groupByNames: groupBy,
+      source: "BATCH",
+    }
+    AnalyticsUtils.filterBody(filterBodyEntity)
+  }
+  let fetchFilterData = async dimensions => {
+    try {
+      let groupBy = getStringListFromArrayDict(dimensions)
+      let filterUri = `${Window.env.apiBaseUrl}/analytics/v1/filters/${domain}`
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/PerformanceMonitor/PerformanceMonitor.res:37]: Can we move this Url to APIUtils ?

File: src/screens/Analytics/PerformanceMonitor/PieChartPerformance/PieChartPerformanceUtils.res [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        },
+        "showInLegend": true,
+      },
+    },
+    "legend": {
+      "align": "right", // Align the legend to the right
+      "verticalAlign": "middle", // Vertically center the legend
+      "layout": "vertical", // Use a vertical layout for legend items
+      "width": "35%",
+      // "y": 35,
```
Comment (Riddhiagrawal001) [inline src/screens/Analytics/PerformanceMonitor/PieChartPerformance/PieChartPerformanceUtils.res:36]: If we don't need this can we remove it

**Comment by JeevaRamu0104** [comment]
> * Please add Title and subtitle to the page
> * Borders to the charts to maintain consistency with analytics of other pages
> * Right align the cards at the same line as the 2nd row is overflowing a bit
> * Gap consistency
> * Section heading (to be discussed)

This will be addressed by @sagarnaikjuspay

**Review by Riddhiagrawal001 — APPROVED** [review]

**Review by PritishBudhiraja — APPROVED** [review]

**Pull request merged**